api/service.proto

```diff
@@ -91,61 +91,70 @@ message DeleteArtifactRequest {
 
 
 service MetadataService {
-  rpc CreateArtifactType(CreateArtifactTypeRequest)
-      returns (CreateArtifactTypeResponse) {
+
+  // NOTE:
+  // The order of the following RPC methods affects the order of matching
+  // a particular HTTP path. So put a more specific path pattern before
+  // a generic one. For example,
+  // GET /api/v1alpha1/{parent=artifact_types/**}/artifacts
+  // should appear before
+  // GET /api/v1alpha1/{name=artifact_types/**} to be possibly matched.
+
+  rpc CreateArtifact(CreateArtifactRequest)
+      returns (CreateArtifactResponse) {
     option (google.api.http) = {
-      post: "/api/v1alpha1/artifact_types"
-      body: "artifact_type"
+      post: "/api/v1alpha1/{parent=artifact_types/**}/artifacts"
+      body: "artifact"
     };
   }
 
-  rpc GetArtifactType(GetArtifactTypeRequest)
-      returns (GetArtifactTypeResponse) {
+  rpc GetArtifact(GetArtifactRequest)
+      returns (GetArtifactResponse) {
     option (google.api.http) = {
-      get: "/api/v1alpha1/{name=artifact_types/*}"
+      get: "/api/v1alpha1/{name=artifact_types/**/artifacts/*}"
     };
   }
 
-  rpc ListArtifactTypes(ListArtifactTypesRequest)
-      returns (ListArtifactTypesResponse) {
+  rpc ListArtifacts(ListArtifactsRequest)
+      returns (ListArtifactsResponse) {
     option (google.api.http) = {
-      get: "/api/v1alpha1/artifact_types"
+      get: "/api/v1alpha1/{name=artifact_types/**}/artifacts"
     };
   }
 
-  rpc DeleteArtifactType(DeleteArtifactTypeRequest)
+  rpc DeleteArtifact(DeleteArtifactRequest)
       returns (google.protobuf.Empty) {
     option (google.api.http) = {
-      delete: "/api/v1alpha1/{name=artifact_types/*}"
+      delete: "/api/v1alpha1/{name=artifact_types/**/artifacts/*}"
     };
   }
 
-  rpc CreateArtifact(CreateArtifactRequest)
-      returns (CreateArtifactResponse) {
+  rpc CreateArtifactType(CreateArtifactTypeRequest)
+      returns (CreateArtifactTypeResponse) {
     option (google.api.http) = {
-      post: "/api/v1alpha1/{parent=artifact_types/*}/artifacts"
-      body: "artifact"
+      post: "/api/v1alpha1/artifact_types"
+      body: "artifact_type"
     };
   }
 
-  rpc GetArtifact(GetArtifactRequest)
-      returns (GetArtifactResponse) {
+  rpc GetArtifactType(GetArtifactTypeRequest)
+      returns (GetArtifactTypeResponse) {
     option (google.api.http) = {
-      get: "/api/v1alpha1/{name=artifact_types/*/artifacts/*}"
+      get: "/api/v1alpha1/{name=artifact_types/**}"
     };
   }
 
-  rpc ListArtifacts(ListArtifactsRequest)
-      returns (ListArtifactsResponse) {
+  rpc ListArtifactTypes(ListArtifactTypesRequest)
```

Comment:
Based on your note above, you should set `ListArtifactTypes` before `GetArtifactType`, otherwise your are not able to call `ListArtifactTypes`. An other opportunity is to rename the url of `GetArtifactType` to /api/v1alpha1/{name=**artifact_type**/**} ,because you're only expecting one instance.

Reply:
Thanks for the review!

`ListArtifactTypes` doesn't cover the pattern in `GetArtifactType`, because `*` matches exactly one path segment.

Also we are following the design guideline [here](https://cloud.google.com/apis/design/resource_names). So all resource prefix must be plural.

Reply:
You are correct, but `GetArtifactType` covers `ListArtifactTypes`, because `**` machtes zero or more URL path segments.  
So if I try to call `ListArtrifactTypes` (`curl -X GET http://localhost:8080/api/v1alpha1/artifact_types`) the response is 
```
{
    "error": "Cannot find type:  SELECT `id`, `name`, `input_type`, `output_type`  from `Type`  WHERE name = '' and is_artifact_type = 1; ",
    "message": "Cannot find type:  SELECT `id`, `name`, `input_type`, `output_type`  from `Type`  WHERE name = '' and is_artifact_type = 1; ",
    "code": 2
}
```   
instead of `{}`, because `GetArtifactType` is invoked. 
By the way I tried to solve the `{}`problem unfortunately ml-metedata do not provide a method to get all artifact types.

Reply:
You're right. I reordered them. Thanks!

> By the way I tried to solve the {}problem unfortunately ml-metedata do not provide a method to get all artifact types.

Yes, it has to be solve on the ml-metadata side.